### PR TITLE
Onboard explicit sharding to deepseek [split version]

### PR DESCRIPTION
**Pull request by NuojCheng**
# Description

This PR introduces explicit sharding support for the DeepSeek model in MaxText. This allows for more granular control over how tensors are distributed across devices, which may lead to better performance and scalability.

This PR is a simplified version based on #2579 .

## Key Changes:

*   **Enabled DeepSeek for Explicit Sharding:** The `deepseek` decoder is now supported in `ShardMode.EXPLICIT`.
*   **Introduced `shard_mode`:** A `shard_mode` parameter has been added to various layers, including attention, embeddings, and MoE, to control sharding behavior. This allows for more flexible and explicit sharding configurations.
*   **Refactored Sharding Logic:** The existing sharding logic has been (minorly) refactored to use the new sharding utilities and `NamedSharding` objects, making the sharding implementation more explicit and maintainable.
*   **Updated Tests:** The test suite has been updated to include tests for explicit sharding, ensuring the correctness and robustness of the implementation.

## Detailed Changes by File:

*   **`src/MaxText/common_types.py`**: Added `Q_LORA_UP_PROJ` and `KV_LORA_UP_PROJ` to support LoRA projections with explicit sharding.
*   **`src/MaxText/configs/types.py`**: Enabled the `deepseek` decoder for explicit sharding.
*   **`src/MaxText/layers/attention_mla.py`**: Updated to use `out_sharding` and `_maybe_shard_with_logical` for more explicit control over sharding in MLA.
*   **`src/MaxText/layers/attentions.py`**: Added the `shard_mode` parameter to `RotaryEmbedding` and its variants. Refactored input and projection sharding to be more explicit.
*   **`src/MaxText/layers/deepseek.py`**: Integrated explicit sharding into the `DeepSeekDecoderLayer` by using `_maybe_shard_with_logical` and passing `out_sharding` and `intermediate_sharding` to sub-layers.
*   **`src/MaxText/layers/embeddings.py`**: Added the `shard_mode` parameter to all embedding classes to allow for explicit sharding configuration.
*   **`src/MaxText/layers/moe.py`**: Added `shard_mode` to `GateLogit` and updated the `MoeBlock` to handle explicit sharding for weights and activations.
*   **`src/MaxText/sharding.py`**: Introduced new utility functions for explicit sharding.
*   **`tests/*`**: Updated various tests to include test cases for explicit sharding and to pass the new `shard_mode` and `mesh` parameters where necessary.

# Tests

Model_name: deepseek3-test
Topology: v5p-8
JAX==0.8.1
TL;DR: 
* Auto Sharding: Maintains performance parity with the main branch (deviations ≤ 0.3%).
* Explicit Sharding: While this introduces a slight expected overhead in TP case, it provides significant improvements for TP_transpose sharding.

## Performance Impact Table

| Configuration | Main Peak Memory (Mb) | Main Step Time (ms) | Auto Peak Memory (Mb) | Auto Peak Mem Change (%) | Auto Step Time (ms) | Auto Step Time Change (%) | Explicit Peak Memory (Mb) | Explicit Peak Mem Change (%) | Explicit Step Time (ms) | Explicit Step Time Change (%) |
| :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- |
| **FSDP4** | 68748 | 2238 | 68748 | *0%* | 2238 | *0%* | 68961 | *0.31%* | 2284 | *2.01%* |
| **FSDP2+TP2** | 69447 | 2009 | 69448 | *0%* | 2014 | *0.25%* | 69617 | *0.24%* | 2378 | *18.37%* |
| **FSDP2+TPT2** | 93835 | 10742 | 93836 | *0%* | 10743 | *0.01%* | 69373 | **-26.07%** | 2714 | **-74.68%** |

## Full FSDP
```
smoke_train model_name=deepseek3-test per_device_batch_size=1
```
* 0ea9b55e4cff49137fcd14e72dfdfb6ba6243e02: [xprof](http://xprof.corp.google.com/trace_viewer/chengnuojin-14798487102806104965)
* `shard_mode=auto`: [xprof](http://xprof.corp.google.com/trace_viewer/chengnuojin-6789662119985341136)
* `shard_mode=explicit`: [xprof](http://xprof.corp.google.com/trace_viewer/chengnuojin-18266912542486481876)

## FSDP + TP
```
smoke_train model_name=deepseek3-test per_device_batch_size=1 ici_tensor_parallelism=2
```

* 0ea9b55e4cff49137fcd14e72dfdfb6ba6243e02: [xprof](http://xprof.corp.google.com/trace_viewer/chengnuojin-1325523371827821617)
* `shard_mode=auto`: [xprof](http://xprof.corp.google.com/trace_viewer/chengnuojin-12218672606251075329)
* `shard_mode=explicit`: [xprof](http://xprof.corp.google.com/trace_viewer/chengnuojin-1418103694342208187)

## FSDP + TP_transpose
```
smoke_train model_name=deepseek3-test per_device_batch_size=1 ici_tensor_transpose_parallelism=2
```

* 0ea9b55e4cff49137fcd14e72dfdfb6ba6243e02: [xprof](http://xprof.corp.google.com/trace_viewer/chengnuojin-7295329251717598060)
* `shard_mode=auto`: [xprof](http://xprof.corp.google.com/trace_viewer/chengnuojin-12612570281133829429)
* `shard_mode=explicit`: [xprof](http://xprof.corp.google.com/trace_viewer/chengnuojin-15676918277719731237)

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

**Review by gobbleturk — APPROVED** [review]

**Review by RissyRan — COMMENTED** [review]
Thanks for simplifying it! Are you planning to merge this one instead?

If you could directly apply a few minor moments from old PR (just published). We should be good to go.

**Comment by NuojCheng** [comment]
Reply to @RissyRan:

> Are you planning to merge this one instead?

Yes let's merge this one instead. The sharding refactoring parts are removed here.

> If you could directly apply a few minor moments from old PR (just published). We should be good to go.

Thank you for the comments! The concerns from the previous PR are all related to the sharding refactoring, which are not included in this PR.

File: src/MaxText/layers/attention_mla.py [hunk 1]
```diff
@@ -506,47 +525,49 @@ def mla_query_projection(self, inputs_q: Array, inputs_positions: Array, model_m
...
+    q_nope = self._maybe_shard_with_logical(q_nope, query_logical_name)
     q_pe = self.apply_rotary_embedding(q_pe, inputs_positions=inputs_positions)
+    q_pe = self._maybe_shard_with_logical(q_pe, query_logical_name)
     # Query projection is scaled by self.softmax_scale to be consistent MaxText implementation.
     # DeepSeek v3 was doing it in attention score computation.
     query = jnp.concatenate([q_nope, q_pe], axis=-1) * self.softmax_scale
+    query = self._maybe_shard_with_logical(query, query_logical_name)
+    return query
 
+  def mla_get_key_value(self, low_rank_main, key_rope, model_mode):
```
[the remark this answers is not in the document]
Comment (NuojCheng) [reply src/MaxText/layers/attention_mla.py:546]: I don't think the `return query` in line 544 is in an if-else statement. It is a little bit misleading.

File: src/MaxText/layers/attentions.py [hunk 2]
```diff
@@ -985,28 +993,25 @@ def __call__(
...
     """
     if model_mode == MODEL_MODE_PREFILL:
-      inputs_q = self._maybe_shard_with_logical(inputs_q, self.prefill_input_axis_names)
-      inputs_kv = self._maybe_shard_with_logical(inputs_kv, self.prefill_input_axis_names)
+      input_axis_names = self.prefill_input_axis_names
     elif model_mode == MODEL_MODE_TRAIN and self.config.expert_shard_attention_option == EP_AS_CONTEXT:
-      inputs_q = self._maybe_shard_with_logical(inputs_q, self.ep_input_axis_names)
-      inputs_kv = self._maybe_shard_with_logical(inputs_kv, self.ep_input_axis_names)
+      input_axis_names = self.ep_input_axis_names
     elif model_mode == MODEL_MODE_TRAIN:
```
[the remark this answers is not in the document]
Comment (NuojCheng) [reply src/MaxText/layers/attentions.py:999]: Good catch! Updated.

File: src/MaxText/layers/embeddings.py [hunk 3]
```diff
@@ -846,7 +876,14 @@ def __call__(self, inputs: Array, position: None | Array = None) -> Array:
 
     inputs_complex = first_half + 1j * second_half  # shape: [B, S, N, half_dim]
     # Apply the rotary transformation via complex multiplication.
-    rotated = inputs_complex * freqs  # shape: [B, S, N, half_dim]
+    rotated_sharding = (
+        NamedSharding(self.mesh, nn.logical_to_mesh_axes(("activation_batch", "activation_length_no_exp", None, None)))
+        if self.shard_mode == ShardMode.EXPLICIT
+        else None
```
[the remark this answers is not in the document]
Comment (NuojCheng) [reply src/MaxText/layers/embeddings.py:882]: Good idea. I will do `jnp.broadcast_to` and `jnp.multiply` to replace the `jnp.einsum`.
